qemu/tests/block_resize.py

```diff
@@ -1,171 +1,148 @@
-
-    if params.get("format_disk", "no") == "yes":
-        error_context.context("Format disk", logging.info)
-        utils_misc.format_windows_disk(session, params["disk_index"],
-                                       mountpoint=params["disk_letter"])
-
-    disk_update_cmd = params.get("disk_update_cmd")
-    if disk_update_cmd:
-        disk_update_cmd = disk_update_cmd.split("::")
-
-    disk_rescan_cmd = params.get("disk_rescan_cmd")
-
-    block_size = data_image_size
-    disk_change_ratio = params["disk_change_ratio"]
-    for index, ratio in enumerate(disk_change_ratio.strip().split()):
-        old_block_size = block_size
-        block_size = int(int(data_image_size) * float(ratio))
-
-        if block_size == old_block_size:
-            logging.warn("Block size is not changed in round %s."
-                         " Just skip it" % index)
-            continue
-
-        if disk_update_cmd:
-            if "DISK_CHANGE_SIZE" in disk_update_cmd[index]:
-                disk_unit = params.get("disk_unit", "M")
-                size = abs(block_size - old_block_size)
-                change_size = utils_misc.normalize_data_size("%sB" % size,
-                                                             disk_unit)
-                disk_update_cmd[index] = re.sub("DISK_CHANGE_SIZE",
-                                                change_size.split(".")[0],
-                                                disk_update_cmd[index])
-
-        # So far only virtio drivers support online auto block size change in
-        # linux guest. So we need manully update the the disk or even reboot
-        # guest to get the right block size after change it from monitor.
+            disk = utils_disk.get_windows_disks_index(session, img_size)[0]
+            utils_disk.update_windows_disk_attributes(session, disk)
+        error_context.context("Formatting disk", logging.info)
+        mpoint = utils_disk.configure_empty_disk(session, disk, img_size,
+                                                 os_type, fstype=fstype,
+                                                 labeltype=labeltype)[0]
+        partition = mpoint.replace('mnt', 'dev') if 'mnt' in mpoint else None
+
+    for ratio in params.objects("disk_change_ratio"):
+        block_size = int(int(block_virtual_size) * float(ratio))
 
         # We need shrink the disk in guest first, than in monitor
-        if block_size < old_block_size and disk_update_cmd:
+        if float(ratio) == 0.9:
             error_context.context("Shrink disk size to %s in guest"
                                   % block_size, logging.info)
-            session.cmd(disk_update_cmd[index])
+            if os_type == 'windows':
+                shr_size = utils_misc.normalize_data_size(str(
+                    utils_disk.get_disk_size(session, os_type, disk) -
+                    block_size), order_magnitude='M').split(".")[0]
+                drive.shrink_volume(session, mpoint, shr_size)
```

Comment:
I think it still shrinks the system disk letter `C `not for the data disk letter when run case `with_data_disk`.

Reply:
for system disk, the mpoint = C, for data disk, the mpoint = new_letter(return by 84 line), won't hit your said situation.